Objects/weakrefobject.c

```diff
@@ -48,6 +128,59 @@ new_weakref(PyObject *ob, PyObject *callback)
     return result;
 }
 
+#ifdef Py_GIL_DISABLED
+
+// Clear the weakref and steal its callback into `callback`, if provided.
+static void
+clear_weakref_lock_held(PyWeakReference *self, PyObject **callback)
+{
+    // TODO: Assert locks are held or world is stopped
+    if (_Py_atomic_load_ptr(&self->wr_object) != Py_None) {
+        PyWeakReference **list = GET_WEAKREFS_LISTPTR(self->wr_object);
+        if (*list == self)
+            /* If 'self' is the end of the list (and thus self->wr_next ==
+               NULL) then the weakref list itself (and thus the value of *list)
+               will end up being set to NULL. */
+            _Py_atomic_store_ptr(list, self->wr_next);
+        _Py_atomic_store_ptr(&self->wr_object, Py_None);
+        if (self->wr_prev != NULL)
+            self->wr_prev->wr_next = self->wr_next;
+        if (self->wr_next != NULL)
+            self->wr_next->wr_prev = self->wr_prev;
+        self->wr_prev = NULL;
+        self->wr_next = NULL;
+    }
+    if (callback != NULL) {
+        *callback = self->wr_callback;
+        self->wr_callback = NULL;
+    }
+}
+
+// Clear the weakref while the world is stopped
+static void
+gc_clear_weakref(PyWeakReference *self)
```

Comment:
Can these functions (`clear_weakref_lock_held`, `gc_clear_weakref`, `clear_weakref`) be combined with their default build implementations?